substrate-processor/src/batch.ts

```diff
@@ -6,7 +6,16 @@ import {QualifiedName} from "./interfaces/substrate"
 import {Heap} from "./util/heap"
 import {Range, rangeDifference, rangeIntersection} from "./util/range"
 
-
+/**
+ * Collects handlers for the various trigger types (pre/post block, event, extrinsic, EVM log).
+ * Keeps a list of handlers of the type corresponding to the trigger
+ * 
+ * @property pre: array of {@link BlockHandler}
```

Comment:
The way you are proposing does not show up in an IDE when hovering over the class. It does not allow people to discover internal properties.

<img width="483" alt="image" src="https://user-images.githubusercontent.com/6950554/163324123-cfdf5400-8162-4a5a-af89-1d04fb46b245.png">

With `@properties`, they are visible from the class or interface. Since self-discovery is one of the main issues of our users...

Reply:
One is not supposed to see all class properties just by hovering over class or interface. Usually this is just too much.
Developers typically explore stuff via "jump to definition".

There is also another way:

<img width="720" alt="image" src="https://user-images.githubusercontent.com/762157/163327911-de17f384-45f6-41c9-93d0-edf20339a440.png">

<img width="946" alt="image" src="https://user-images.githubusercontent.com/762157/163328049-c29eea7b-2168-43cb-89a6-ed63b04cd11c.png">

Reply:
The main issue we want to address is to show the developers all possible options when developing mappings. It's virtually impossible to figure out that there options like 'setBatchSize' or 'triggerEvents'. IDE should help with that, so whenever you type 'processor...' all possible setters are visible with a short description. 

Same for handlers and context definitions -- the main goal is to make sure the developers are aware what is possible with Subsquid. The details can be then figured out via external docs or through the code (depending on how vigilant they are)

Reply:
<img width="946" alt="image" src="https://user-images.githubusercontent.com/762157/163342893-ea302529-29c9-4898-8dd1-a6b0d05c3df2.png">

Reply:
> Developers typically explore stuff via "jump to definition".

I know I do, that's what I did to understand the project and write the docs, but judging on the question we get on a daily basis, not many others do. Unfortunately. 😢 

I understand what you are saying Eldar, the thing is that having Interfaces as function arguments "hides" what's inside it.

<img width="462" alt="image" src="https://user-images.githubusercontent.com/6950554/163599665-93e60193-16b5-474b-ba58-321200693133.png">

If we want people to self serve and do the documentation as you suggest (not that I am against it, in principle you are absolutely correct), they need to know what Range is, what it looks like. Ultimately, what they need to put as an argument to the function. (It's the same thing for `setDataSource`, by the way).

But I looked at the comment of the function itself, while writing this comment:

```
    /**
     * Sets a range of blocks to restrict processing of a blockchain. (optional)
     * When used, the processor will start extracting information from the specified `from` block in the {@link Range} 
     * object and (optionally) stop at the `to` indicated block.
     * 
     * If not set, the processor will default to start from block number 0 and never stop.
     * 
     * @param range an object with `from` and (optionally) `to` fields, defining the block numbers to which the chain 
     * processing should be restricted. @see Range
     */
    setBlockRange(range: Range): void {
        this.assertNotRunning()
        this.blockRange = range
    }
```

And it changed my mind a little bit. I tried to do it in one go, because I wanted to get other things done and probably by the end of it I was not lucid.

I am going to do another pass and try to make sure that functions are accessible and easy to use.

About what Dimitry said: my suggestion is that if methods like `setDataSource` are mandatory, they should not be "discovered" by doing `processor.` and waiting for the IDE to suggest it. Usually, mandatory settings are part of the constructor, or there are patterns like [Factory Method](https://en.wikipedia.org/wiki/Factory_method_pattern). You could even have what in Python are "class methods" that call the constructor to initialize parameters, like:

```typescript

export class SubstrateProcessor {

  getProcessorFromDataSource(name: string, src: DataSource): SubstrateProcessor {
    const processor = new SubstrateProcessor(name);
    processor.setDataSource(src);
    return processor;
  } 
}
```

Silly quick example, but I hope you get the idea.